expect toMatchObject should show which object properties caused the assertion to fail

When using expect(...).toMatchObject with complex objects, it's hard to see which property/properties caused the assertion to fail.

**Steps to Reproduce**

```ts
import { it } from '@std/testing/bdd';
import { expect } from '@std/expect';

const expected = {
  field1: 'abc',
  field2: 'def',
  foo: 'bar',
}

it('toMatchObject', () => {
  expect({ 
    field1: 'abc',
    field2: 'def',
    foo: 'baz',
  }).toMatchObject(expected);
});
```

```
$ deno test denobug.test.ts 
running 1 test from ./denobug.test.ts
toMatchObject ... FAILED (2ms)

 ERRORS 

toMatchObject => https://jsr.io/@std/testing/1.0.5/_test_suite.ts:218:10
error: AssertionError: Expected {
  field1: "abc",
  field2: "def",
  foo: "baz",
} to NOT match {
  field1: "abc",
  field2: "def",
  foo: "bar",
}
```

**Describe the solution you'd like**

The library could indicate which property did not match, at lease for the top-level properties (in this case, `foo`).

**Describe alternatives you've considered**

`assertObjectMatch` and Jest do this, for all properties, including nested ones:

![image](https://github.com/user-attachments/assets/0f78429b-2b92-4194-b283-53d2a87ca94a)

But obviously, `assertObjectMatch` doesn't support matchers, so there's no easy workaround for this issue.

--- a/expect/_build_message.ts
+++ b/expect/_build_message.ts
@@ -18,9 +18,9 @@ function isString(value: unknown): value is string {
 export function buildEqualErrorMessage<T>(
   actual: T,
   expected: T,
-  options: EqualErrorMessageOptions,
+  options: EqualErrorMessageOptions = {},
 ): string {
-  const { formatter = format, msg } = options ?? {};
+  const { formatter = format, msg } = options;
   const msgPrefix = msg ? `${msg}: ` : "";
   const actualString = formatter(actual);
   const expectedString = formatter(expected);
@@ -40,9 +40,9 @@ export function buildEqualErrorMessage<T>(
 export function buildNotEqualErrorMessage<T>(
   actual: T,
   expected: T,
-  options: EqualErrorMessageOptions,
+  options: EqualErrorMessageOptions = {},
 ): string {
-  const { msg } = options ?? {};
+  const { msg } = options;
   const actualString = String(actual);
   const expectedString = String(expected);
 

--- a/expect/_build_message.ts
+++ b/expect/_build_message.ts
@@ -42,9 +42,9 @@ export function buildNotEqualErrorMessage<T>(
   expected: T,
   options: EqualErrorMessageOptions = {},
 ): string {
-  const { msg } = options;
-  const actualString = String(actual);
-  const expectedString = String(expected);
+  const { formatter = format, msg } = options;
+  const actualString = formatter(actual);
+  const expectedString = formatter(expected);
 
   const msgPrefix = msg ? `${msg}: ` : "";
   return `${msgPrefix}Expected actual: ${actualString} not to be: ${expectedString}.`;

--- a/expect/_matchers.ts
+++ b/expect/_matchers.ts
@@ -21,6 +21,10 @@ import {
   iterableEquality,
   subsetEquality,
 } from "./_utils.ts";
+import {
+  buildEqualErrorMessage,
+  buildNotEqualErrorMessage,
+} from "./_build_message.ts";
 
 export function toBe(context: MatcherContext, expect: unknown): MatchResult {
   if (context.isNot) {
@@ -565,7 +569,7 @@ export function toMatchObject(
     );
   }
 
-  const pass = equal(context.value, expected, {
+  const pass = equal(received, expected, {
     strictCheck: false,
     customTesters: [
       ...context.customTesters,
@@ -575,20 +579,15 @@ export function toMatchObject(
   });
 
   const triggerError = () => {
-    const actualString = format(context.value);
-    const expectedString = format(expected);
-
     if (context.isNot) {
-      const defaultMessage =
-        `Expected ${actualString} to NOT match ${expectedString}`;
+      const defaultMessage = buildNotEqualErrorMessage(received, expected);
       throw new AssertionError(
         context.customMessage
           ? `${context.customMessage}: ${defaultMessage}`
           : defaultMessage,
       );
     } else {
-      const defaultMessage =
-        `Expected ${actualString} to match ${expectedString}`;
+      const defaultMessage = buildEqualErrorMessage(received, expected);
       throw new AssertionError(
         context.customMessage
           ? `${context.customMessage}: ${defaultMessage}`

--- a/expect/_to_match_object_test.ts
+++ b/expect/_to_match_object_test.ts
@@ -119,16 +119,16 @@ Deno.test("expect().toMatchObject() throws the correct error messages", () => {
   {
     const e = assertThrows(
       () => expect({ a: 1 }).toMatchObject({ a: 2 }),
-      Error,
+      AssertionError,
     );
-    assertNotMatch(e.message, /NOT/);
+    assertNotMatch(e.message, /not to be/);
   }
   {
     const e = assertThrows(
       () => expect({ a: 1 }).not.toMatchObject({ a: 1 }),
-      Error,
+      AssertionError,
     );
-    assertMatch(e.message, /NOT/);
+    assertMatch(e.message, /not to be/);
   }
 });
 

--- a/expect/_to_match_object_test.ts
+++ b/expect/_to_match_object_test.ts
@@ -65,7 +65,7 @@ Deno.test("expect().toMatchObject() with array", () => {
   ]);
 });
 
-Deno.test("expect(),toMatchObject() with asyAsymmetric matcher", () => {
+Deno.test("expect().toMatchObject() with asymmetric matcher", () => {
   expect({ position: { x: 0, y: 0 } }).toMatchObject({
     position: {
       x: expect.any(Number),
@@ -130,4 +130,101 @@ Deno.test("expect().toMatchObject() throws the correct error messages", () => {
     );
     assertMatch(e.message, /NOT/);
   }
+});
+
+Deno.test("expect().toMatchObject() displays a diff", async (t) => {
+  await t.step("with expect.any", () => {
+    assertThrows(
+      () =>
+        expect({ a: "a" })
+          .toMatchObject({ a: expect.any(Number) }),
+      AssertionError,
+      `    {
++     a: Any {
++       inverse: false,
++       value: [Function: Number],
++     },
+-     a: "a",
+    }`,
+    );
+  });
+
+  await t.step("with nested properties", () => {
+    const x = {
+      command: "error",
+      payload: {
+        message: "NodeNotFound",
+      },
+      protocol: "graph",
+    };
+
+    const y = {
+      protocol: "graph",
+      command: "addgroup",
+      payload: {
+        graph: "foo",
+        metadata: {
+          description: "foo",
+        },
+        name: "somegroup",
+        nodes: [
+          "somenode",
+          "someothernode",
+        ],
+      },
+    };
+
+    assertThrows(
+      () => expect(x).toMatchObject(y),
+      AssertionError,
+      `    {
++     command: "addgroup",
+-     command: "error",
+      payload: {
++       graph: "foo",
++       metadata: {
++         description: "foo",
++       },
++       name: "somegroup",
++       nodes: [
++         "somenode",
++         "someothernode",
++       ],
+-       message: "NodeNotFound",
+      },
+      protocol: "graph",
+    }`,
+    );
+
+    assertThrows(
+      () => expect({ foo: [] }).toMatchObject({ foo: ["bar"] }),
+      AssertionError,
+      `    {
++     foo: [
++       "bar",
++     ],
+-     foo: [],
+    }`,
+    );
+  });
+
+  await t.step(
+    "with `__proto__`",
+    () => {
+      const objectA = { ["__proto__"]: { polluted: true } };
+      const objectB = { ["__proto__"]: { polluted: true } };
+      const objectC = { ["__proto__"]: { polluted: false } };
+      expect(objectA).toMatchObject(objectB);
+      assertThrows(
+        () => expect(objectA).toMatchObject(objectC),
+        AssertionError,
+        `    {
+      ['__proto__']: {
+-       polluted: true,
++       polluted: false,
+      },
+    }`,
+      );
+    },
+  );
 });

--- a/internal/format_test.ts
+++ b/internal/format_test.ts
@@ -27,7 +27,7 @@ Deno.test("format() overrides default behaviours of Deno.inspect", async (t) => 
   // Wraps objects into multiple lines even when they are small. Prints trailing
   // commas.
   await t.step(
-    "fromat() always wraps objects into multiple lines and prints trailing commas",
+    "format() always wraps objects into multiple lines and prints trailing commas",
     () =>
       assertEquals(
         stripAnsiCode(format({ a: 1, b: 2 })),
@@ -37,6 +37,12 @@ Deno.test("format() overrides default behaviours of Deno.inspect", async (t) => 
 }`,
       ),
   );
+
+  await t.step("format() sorts properties", () =>
+    assertEquals(
+      format({ b: 2, a: 1 }),
+      format({ a: 1, b: 2 }),
+    ));
 
   await t.step("format() wraps Object with getters", () =>
     assertEquals(

